dotnet/src/Connectors/Connectors.Memory.Postgres/PostgresVectorStoreRecordCollection.cs

```diff
@@ -0,0 +1,291 @@
+        VectorStoreRecordPropertyVerification.VerifyGenericDataModelDefinitionSupplied(typeof(TRecord), options?.VectorStoreRecordDefinition is not null);
+
+        // Assign.
+        this._client = client;
+        this.CollectionName = collectionName;
+        this._options = options ?? new PostgresVectorStoreRecordCollectionOptions<TRecord>();
+        this._propertyReader = new VectorStoreRecordPropertyReader(
+            typeof(TRecord),
+            this._options.VectorStoreRecordDefinition,
+            new()
+            {
+                RequiresAtLeastOneVector = false,
+                SupportsMultipleKeys = false,
+                SupportsMultipleVectors = true,
+            });
+
+        // Validate property types.
+        this._propertyReader.VerifyKeyProperties(PostgresConstants.SupportedKeyTypes);
+        this._propertyReader.VerifyDataProperties(PostgresConstants.SupportedDataTypes, supportEnumerable: true);
+        this._propertyReader.VerifyVectorProperties(PostgresConstants.SupportedVectorTypes);
+
+        // Resolve mapper.
+        // First, if someone has provided a custom mapper, use that.
+        // If they didn't provide a custom mapper, and the record type is the generic data model, use the built in mapper for that.
+        // Otherwise, don't set the mapper, and we'll default to just using Azure AI Search's built in json serialization and deserialization.
+        if (this._options.DictionaryCustomMapper is not null)
+        {
+            this._mapper = this._options.DictionaryCustomMapper;
+        }
+        else if (typeof(TRecord).IsGenericType && typeof(TRecord).GetGenericTypeDefinition() == typeof(VectorStoreGenericDataModel<>))
+        {
+            this._mapper = (new PostgresGenericDataModelMapper<TKey>(this._propertyReader) as IVectorStoreRecordMapper<TRecord, Dictionary<string, object?>>)!;
+        }
+        else
+        {
+            this._mapper = new PostgresVectorStoreRecordMapper<TRecord>(this._propertyReader);
+        }
+    }
+
+    /// <inheritdoc/>
+    public async Task<bool> CollectionExistsAsync(CancellationToken cancellationToken = default)
+    {
+        return await this._client.DoesTableExistsAsync(this.CollectionName, cancellationToken).ConfigureAwait(false);
+    }
+
+    /// <inheritdoc/>
+    public async Task CreateCollectionAsync(CancellationToken cancellationToken = default)
+    {
+        await this._client.CreateTableAsync(this.CollectionName, this._propertyReader.RecordDefinition, false, cancellationToken).ConfigureAwait(false);
+        // Create indexes for vector properties.
+        foreach (var vectorProperty in this._propertyReader.VectorProperties)
+        {
+            await this._client.CreateVectorIndexAsync(this.CollectionName, vectorProperty, cancellationToken).ConfigureAwait(false);
```

Comment:
Maybe the vector type and max dimensions need to be checked to decide whether to create an index or not?

> - vector - up to 2,000 dimensions
> - halfvec - up to 4,000 dimensions (added in 0.7.0)
> - bit - up to 64,000 dimensions (added in 0.7.0)
> - sparsevec - up to 1,000 non-zero elements (added in 0.7.0)

Reply:
Good call. I'll put a check in CreateCollectionAsync that tests for this case and throws if the index can't be created with the given dimensions.

Worth noting that halfvec, bit, and sparsevec are not supported. VectorStoreRecordVectorProperty doesn't have a way to differentiate the vector type, just to set the index and distance. I think implementing different vector types, along with the ability to query indexes that are built e.g. on halfvec on a vector column, and requiring a query like [`SELECT * FROM items ORDER BY embedding::halfvec(3) <-> '[1,2,3]' LIMIT 5;`](https://github.com/pgvector/pgvector?tab=readme-ov-file#half-precision-indexing), are going to be difficult to support and will require more thought.

Reply:
Maybe just a warning is enough?

No index
> By default, pgvector performs exact nearest neighbor search, which provides perfect recall.

Of course, vector is sufficient for now, and other types can be considered later.

Reply:
Modified to log instead of throwing an exception.